crypto/keyring/keyring.go

```diff
@@ -845,88 +846,104 @@ func (ks keystore) newRecord(name string, pk types.PubKey, item isRecord_Item) (
 		}
 		// key not found, all good: assume version = 0
 	} else {
 		if len(item.Data) != 4 {
-			return migrated, sdkerrors.ErrInvalidVersion.Wrapf(
-				"Can't migrate the keyring - the stored version is malformed: [%v]: %v",
+			return false, sdkerrors.ErrInvalidVersion.Wrapf(
+				"Can't migrate the keyring - the stored version is malformed: [%s]: %s",
 				item.Description, string(item.Data))
 		}
 		version = binary.LittleEndian.Uint32(item.Data)
 	}
-	return ks.migrate(version, item, migrated)
+	return ks.migrate(version, item)
 }
 
-func (ks keystore) migrate(version uint32, i keyring.Item, migrated bool) (bool, error) {
+func (ks keystore) migrate(version uint32, i keyring.Item) (bool, error) {
 	if version == CURRENT_VERSION {
 		// return nil
-		return migrated, fmt.Errorf("versions match")
+		return false, fmt.Errorf("versions match")
 	}
 	if version > CURRENT_VERSION {
-		return migrated, sdkerrors.ErrInvalidVersion.Wrapf(
+		return false, sdkerrors.ErrInvalidVersion.Wrapf(
 			"Can't migrate the keyring - wrong keyring version: [%v]: %v, expected version to be max %d",
 			i.Description, string(i.Data), CURRENT_VERSION)
 	}
+	
+	// TODO i can use map or array of boleeans to check if no error occured thenw eu pdate the version otherwise
+	// return nil or error
+	// 6. at the end of the loop update version
+	migrated, err := ks.performMigration()
+	if !migrated {
+		return migrated, err
+	} else {
+		if err := ks.setVersion(); err != nil {
+			return migrated, err
+		}
+	}
+	
+	return migrated, nil
+}
+
+func (ks keystore) performMigration() (bool, error) {
 	keys, err := ks.db.Keys()
 	if err != nil {
 		// return err
-		return migrated, fmt.Errorf("Keys() error, err: %s", err)
+		return false, fmt.Errorf("Keys() error, err: %s", err)
 	}
 
-	for _, key := range keys {
-		// TODO to move it to separate function that outputs an error
+	migrations := make([]bool, len(keys))
+
+	for i, key := range keys {
 		if !strings.HasSuffix(key, infoSuffix) {
 			fmt.Printf("key %s has no infoSuffix", key)
 			continue
 		}
+
 		item, err := ks.db.Get(key)
 		if err != nil {
-			//return err
-			return migrated, fmt.Errorf("Get error, err - %s", err)
+			return false, fmt.Errorf("Get error, err - %s", err)
 		}
 
 		if len(item.Data) == 0 {
 			// return sdkerrors.Wrap(sdkerrors.ErrKeyNotFound, key)
-			return migrated, fmt.Errorf("sdkerrors.Wrap(sdkerrors.ErrKeyNotFound, key), err: %s", sdkerrors.Wrap(sdkerrors.ErrKeyNotFound, key))
+			return false, fmt.Errorf("sdkerrors.Wrap(sdkerrors.ErrKeyNotFound, key), err: %s", sdkerrors.Wrap(sdkerrors.ErrKeyNotFound, key))
 		}
 
 		// 2.try to deserialize using proto, if good then continue, otherwise try to deserialize using amino
 		if _, err := ks.protoUnmarshalRecord(item.Data); err == nil {
 			fmt.Println("protoUnmarshalRecord continue")
+			//TODO decide if we do not migrate, we should mark it as false or true?
+			migrations[i] = false 
```

Comment:
We have to decide if we do not migrate a key, should we we mark it as true or false?